### DISCUSSION TITLE
How do you create a website with multiple languages?

### DISCUSSION THREAD

**Discussion by markseuffert** [datenstrom/community#14] [Ask a question] [answered]
How do you create a website with multiple languages?

**Comment by markseuffert** [comment]
The previous solutions didn't work. Here are some things that would be good to have:
- locations with language prefix, e.g. `http://website/en/` `http://website/de/`
- optional default language, location without language prefix, e.g. `http://website/`
- optional links between same page in different languages

Anything else you like to have?

**Comment by dcschmid** [comment]
i would prefer the first solution with the optional default language.

**Comment by markseuffert** [comment]
Here's is my wish list for the multi language support:
- Can be enabled in the configuration
- Can be configured via file system, different folders for different languages
- Option to change language, for whole site
- Option to switch language, for same page in different languages
- Templates and snippets work unchanged, maybe additional language snippets
- Possible to have no landing page, instead home page in default language
- Possible to create static files, as usual with support for creating whole site in one go
- Search engine friendly (clean URLs, language meta data)

**Comment by markseuffert** [comment, marked as answer]
I added support for multiple languages.

To enable multi language mode open `system/config/config.ini` and change `multiLanguageMode = 1`. Create a new folder `1-en` in your content directory,  put all your previous content inside this folder. This is now the default language for your website. If your default language is not English, please name the folder correspondingly.

![screen shot 2014-10-07 at 01 29 21](https://cloud.githubusercontent.com/assets/2429165/4535292/de942ec0-4db0-11e4-9c72-8b45dc3ee5be.png)

The screenshot shows the folders `1-en`, `2-de` and `3-fr` for English, German and French. You can add as many languages as you like. If you want a landing page, add a `default` folder as well. The landing page can redirect to the correct language. The API has examples to show all available languages.

As a final step translate text lines in file `system/config/text.ini`, use one file per language.

Let me know how it works. :)

**Comment by dcschmid** [comment]
Hi Mark,

must the text.ini File for each language a special naming and where must be saved?

Thanks! :-)

Daniel

**Comment by markseuffert** [comment]
Hi Daniel, 
the name for text files should be `text*.ini` (\* = any text). For example Englisch, German and French:

`system/config/text.ini`
`system/config/text-deutsch.ini`
`system/config/text-français.ini`

**Comment by dcschmid** [comment]
Hi Mark,

thanks for your support. Now its works and Yellow is a great cms :-)

Daniel

**Comment by markseuffert** [comment]
Thanks for testing, Daniel! Hope it is okay if I ask some questions. Was the configuration difficult? Anything missing that you need for a multi language website?

**Comment by dcschmid** [comment]
Hi Mark,

the configuration was clean and simple. Once is missing, on the second language i have the /en/ in the url, but not in the default language. 

Thanks Daniel

**Comment by markseuffert** [comment]
Yes that's possible, add a default folder (without any files).

![screen shot 2014-10-07 at 22 23 12](https://cloud.githubusercontent.com/assets/2429165/4549664/fabe5e18-4e5f-11e4-9fe3-9cf4703216d5.png)

The screenshot shows the folders for English, German, French and a default folder. The `default` folder creates a landing page, which automatically detects the browser language and redirects to the correct language. This gives you URLs with language prefix, e.g. `http://website/en/` `http://website/de/` `http://website/fr/`.

**Comment by dajare** [comment]
I just added this information to the wiki page, taken from the thread above:

```
system/config/text.ini
system/config/text-deutsch.ini
system/config/text-français.ini
```

But wouldn't it make it easier (shorter, more consistent) to use the same 2-letter locale name as in the folder conventions?

```
system/config/text.ini
system/config/text-de.ini
system/config/text-fr.ini
```

And presumably if you have a default language that is not English, (e.g. `1-de`), then you have an English text.ini (i.e., `system/config/text-en.ini`)? 

And if so (!), does that mean that a `text-en.ini` would take priority over `text.ini`? (That is, you could have a set of default strings in English, but tweak them, still in English, for any reason in a deployment.)

Thanks for this!

**Comment by markseuffert** [comment]
Thank you for the update!

Yes, I agree that a 2-letter name would be more consistent. I suggest to name these files:

```
system/config/language-de.ini
system/config/language-en.ini
system/config/language-fr.ini
```

About a default language, at the moment there isn't one... however the default installation comes with English text strings, which probably makes `language-en.txt` the default. What do you think?

**Comment by markseuffert** [comment]
Hope it's okay, I updated Yellow and wiki page to the new format. Thanks David for the good idea!  :+1:
